### DISCUSSION TITLE
group dependencies in one MR

### DISCUSSION THREAD

**Discussion by ikercrg** [renovatebot/renovate#9804] [Request Help]
"renovateVersion": "25.5.0"

In one MR (all starting @angular/ except @angular/cli):
    "@angular/animations": "10.2.5",
    "@angular/cdk": "10.2.5",
    "@angular/common": "10.2.5",
    "@angular/compiler": "10.2.5",
    "@angular/core": "10.2.5",
    "@angular/forms": "10.2.5",
    "@angular/localize": "10.2.5",
    "@angular/platform-browser": "10.2.5",
    "@angular/platform-browser-dynamic": "10.2.5",
    "@angular/router": "10.2.5",
    "@angular/compiler-cli": "10.2.5"

In one MR:
    "@ngrx/effects": "10.1.2",
    "@ngrx/router-store": "10.1.2",
    "@ngrx/store": "10.1.2",
    "@ngrx/store-devtools": "10.1.2",
    "@ngrx/schematics": "10.1.2",

In one MR:
     "ag-grid-angular": "24.1.0",
    "ag-grid-community": "24.1.0"

In one MR:
    "@angular-eslint/builder": "1.2.0",
    "@angular-eslint/eslint-plugin": "1.2.0",
    "@angular-eslint/eslint-plugin-template": "1.2.0",
    "@angular-eslint/schematics": "1.2.0",
    "@angular-eslint/template-parser": "1.2.0",

If I have this package.json:

```json
"dependencies": {
    "@angular/animations": "10.2.5",
    "@angular/cdk": "10.2.5",
    "@angular/common": "10.2.5",
    "@angular/compiler": "10.2.5",
    "@angular/core": "10.2.5",
    "@angular/forms": "10.2.5",
    "@angular/localize": "10.2.5",
    "@angular/platform-browser": "10.2.5",
    "@angular/platform-browser-dynamic": "10.2.5",
    "@angular/router": "10.2.5",
    "@ng-bootstrap/ng-bootstrap": "7.0.0",
    "@ng-select/ng-select": "5.1.0",
    "@ngrx/effects": "10.1.2",
    "@ngrx/router-store": "10.1.2",
    "@ngrx/store": "10.1.2",
    "@ngrx/store-devtools": "10.1.2",
    "@ngx-translate/core": "13.0.0",
    "@ngx-translate/http-loader": "6.0.0",
    "ag-grid-angular": "24.1.0",
    "ag-grid-community": "24.1.0",
    "bootstrap": "4.6.0",
    "crypto-js": "4.0.0",
    "fbc-iam-angular": "2.0.0",
    "ngx-toastr": "13.2.1",
    "rxjs": "6.6.7",
    "tslib": "2.1.0",
    "zone.js": "0.11.4"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "0.1002.1",
    "@angular-eslint/builder": "1.2.0",
    "@angular-eslint/eslint-plugin": "1.2.0",
    "@angular-eslint/eslint-plugin-template": "1.2.0",
    "@angular-eslint/schematics": "1.2.0",
    "@angular-eslint/template-parser": "1.2.0",
    "@angular/cli": "10.2.3",
    "@angular/compiler-cli": "10.2.5",
    "@foundation/npm-oss-clearing": "2.2.0",
    "@ngrx/schematics": "10.1.2",
    "@types/jest": "24.9.1",
    "@typescript-eslint/eslint-plugin": "4.18.0",
    "@typescript-eslint/parser": "4.18.0",
    "eslint": "7.22.0",
    "eslint-config-airbnb-typescript": "12.3.1",
    "eslint-plugin-cypress": "2.11.2",
    "eslint-plugin-header": "3.1.1",
    "eslint-plugin-import": "2.22.1",
    "eslint-plugin-jest": "24.3.6",
    "eslint-plugin-promise": "4.3.1",
    "eslint-plugin-rxjs": "3.1.3",
    "eslint-plugin-unicorn": "28.0.2",
    "https-proxy-agent": "2.2.4",
    "jest": "24.9.0",
    "jest-marbles": "2.5.1",
    "jest-preset-angular": "7.1.1",
    "jspdf": "2.3.1",
    "ng-mocks": "11.9.1",
    "ngx-translate-testing": "5.1.0",
    "typescript": "4.0.3"
  }
```

and I want to group some dependencies in one MR (in this case, group together the ones with the same name).

Test 1: Fails -> I get one MR with only @angular/cdk upgrade. Here the renovate.json: 

```json
{
  "gradle-lite": {
    "enabled": true
  },
  "gradle": {
    "enabled": false
  },
  "docker-compose": {
    "enabled": false
  },
  "dockerfile": {
    "enabled": false
  },
  "gitlabci": {
    "enabled": false
  },
  "packageRules": [
    {
      "groupName": "angular-eslint",
      "matchPackagePatterns": [
        "^@angular-eslint/"
      ]
    },
    {
      "groupName": "typescript-eslint",
      "matchPackagePatterns": [
        "^@typescript-eslint/"
      ]
    },
    {
      "groupName": "angular",
      "matchPackagePatterns": [
        "^@angular/"
      ],
      "excludePackageNames": [
        "@angular/cli"
      ]
    },
    {
      "groupName": "ngrx",
      "matchPackagePatterns": [
        "^@ngrx/"
      ]
    },
    {
      "groupName": "ag-grid",
      "matchPackagePatterns": [
        "^ag-grid-"
      ]
    },
    {
      "groupName": "spring-boot",
      "matchPackagePrefixes": [
        "org.springframework.boot:"
      ]
    },
    {
      "groupName": "mapstruct",
      "matchPackagePrefixes": [
        "org.mapstruct:"
      ]
    },
    {
      "groupName": "springfox",
      "matchPackagePrefixes": [
        "io.springfox:"
      ]
    }
  ],
  "prConcurrentLimit": 1
}
```

Test 2: Fails -> I get one MR several updates (@angular-devkit/build-angular, @angular/cdk, @typescript-eslint/eslint-plugin, 
@typescript-eslint/parser, cypress, org.apache.tika:tika-parsers, tslib, and much more (npm and gradle dependencies mixed)

```json
{
  "gradle-lite": {
    "enabled": true
  },
  "gradle": {
    "enabled": false
  },
  "docker-compose": {
    "enabled": false
  },
  "dockerfile": {
    "enabled": false
  },
  "gitlabci": {
    "enabled": false
  },
  "packageRules": [
    {
      "groupName": "angular-eslint",
      "matchPackagePrefixes": [
        "@angular-eslint/"
      ]
    },
    {
      "groupName": "typescript-eslint",
      "matchPackagePrefixes": [
        "@typescript-eslint/"
      ]
    },
    {
      "groupName": "core-ui",
      "matchPackagePrefixes": [
        "@siemens/core-ui"
      ],
      "excludePackageNames": [
        "@siemens/core-ui-icons"
      ]
    },
    {
      "groupName": "angular",
      "matchPackagePrefixes": [
        "@angular/"
      ],
      "excludePackageNames": [
        "@angular/cli"
      ]
    },
    {
      "groupName": "ngrx",
      "matchPackagePrefixes": [
        "@ngrx/"
      ]
    },
    {
      "groupName": "ag-grid",
      "matchPackagePrefixes": [
        "ag-grid-"
      ]
    },
    {
      "groupName": "spring-boot",
      "matchPackagePrefixes": [
        "org.springframework.boot:"
      ]
    },
    {
      "groupName": "mapstruct",
      "matchPackagePrefixes": [
        "org.mapstruct:"
      ]
    },
    {
      "groupName": "springfox",
      "matchPackagePrefixes": [
        "io.springfox:"
      ]
    }
  ],
  "prConcurrentLimit": 1
}
```

In the first renovate.json I use matchPackagePatterns and in the second one matchPackagePrefixes, but no one is working. What should I do?

**Comment by viceice** [comment]
Please format your samples and provide your configs.

You should also describe exactly what does should be grouped.

**Reply by ikercrg** [reply]
Sorry, I already  applied the format and I specified it in the beginning how I want to group the dependencies

**Reply by viceice** [reply]
The match patterns are regexes so you need to escape some chars. Eg `\/`

**Reply by viceice** [reply]
Can you copy the config and your package json to a public git repo so we can pr some suggestions. 

You should also check our official group presets for a lot of monorepos. Mostly we use matching by source url.

**Reply by ikercrg** [reply]
You mean something like this https://github.com/ikercrg/test_renovate

**Reply by viceice** [reply]
Yes, thank you. We will look into it. But be patient, it's weekend. 🤗

**Reply by ikercrg** [reply]
jajaja, no problem :) I really appreciate it

**Reply by ikercrg** [reply]
As you can see there are java dependencies too. But I do not think that you need my build.gradle.

Thanks agains

**Reply by ikercrg** [reply]
Hi @viceice, any news related to this?

**Reply by viceice** [reply]
Can your remove the `prConcurrentLimit` from the sample config and enable renovate app on that sample repo, please also extend from `:disableRateLimiting` so we can see all pr's renovate will create.

Can you also add @rarkins and @viceice  as maintainer to your repo(writeaccess), so we can see the logs on https://app.renovatebot.com/dashboard

**Reply by ikercrg** [reply]
@viceice, I think that I have already have done it. Please confirm if it is correct. Thanks

**Reply by viceice** [reply]
Those pr's looks good to me. So what's missing?

**Reply by ikercrg** [reply]
The code says:
```json
{
      "groupName": "typescript-eslint",
      "matchPackagePrefixes": [
        "@typescript-eslint/"
      ]
},
```

and the trace:
DEBUG: Dependency @typescript-eslint/eslint-plugin is part of group angular

The group is not correct

And in one PR I am getting several upgrades, from different groups and so on.

**Reply by viceice** [reply]
Please format code, thanks. Looking into the issue.

**Reply by viceice** [reply]
Can you remove the private deps from package.json, so it will reduce unrelated verbose errors in logs

**Comment by viceice** [comment]
@rarkins I think we have a bug for a rule like this:

```json
    {
        "groupName": "angular",
        "matchPackagePrefixes": [
          "@angular/"
        ],
        "excludePackageNames": [
          "@angular/cli"
        ]
      },
```

It seems to include all deps which are not matching `excludePackageNames`.

@ikercrg Can you reduce the rules to this one only? and reduce the deps to 3:
- `@angular/common`
- `@angular/cli`
- `@typescript-eslint/eslint-plugin`

Those should be enough do demonstrate the bug.

**Reply by viceice** [reply]
ha, i think this is not respecting the prefix match

https://github.com/renovatebot/renovate/blob/b957d4afb7824a165864cbd6019a7540b315ebe9/lib/util/package-rules.ts#L49-L55

**Reply by rarkins** [reply]
Yep this looks like a bug

**Reply by viceice** [reply]
fix is at https://github.com/renovatebot/renovate/pull/9949

**Comment by ikercrg** [comment]
Only 3 devs. Is it correct? Is there a bug?

**Reply by viceice** [reply]
yes, i've found the bug. so please update your repo to only have those 3 deps, it should trigger the bug. so when the new fix is live, we can see if it's really fixed

**Reply by ikercrg** [reply]
```json
"dependencies": {
    "@angular/common": "10.2.5"
  },
  "devDependencies": {
    "@angular/cli": "10.2.3",
    "@typescript-eslint/eslint-plugin": "4.18.0",
  }
```
Only these 3 deps are now in package.json

**Reply by viceice** [reply]
ok, fixed your configs a bit, so now we have the wrong pr's at https://github.com/ikercrg/test_renovate/pull/3 and https://github.com/ikercrg/test_renovate/pull/1

**Reply by ikercrg** [reply]
> ok, fixed your configs a bit, so now we have the wrong pr's at [ikercrg/test_renovate#3](https://github.com/ikercrg/test_renovate/pull/3) and [ikercrg/test_renovate#1](https://github.com/ikercrg/test_renovate/pull/1)

[ikercrg/test_renovate#1](https://github.com/ikercrg/test_renovate/pull/1) here only the title, isn´t it?

**Reply by viceice** [reply]
and the branch name

**Reply by ikercrg** [reply]
I see that it is merged, but I already get bad pr

**Reply by viceice** [reply]
It's not yet live in the app. @rarkins  needs to approve the next update, so it will maybe need a few hours / days

**Reply by ikercrg** [reply]
ok, thanks

**Comment by ikercrg** [comment]
@viceice @rarkins I updated the deps (only angular ones) and I get 3 different Pull Requests. Why do I get this [angular/cdk](https://github.com/ikercrg/test_renovate/pull/18)? Because another upgrade is in [angular major](https://github.com/ikercrg/test_renovate/pull/3) 

I thought that the default behaviour was, one pull request for minor and other one major. But here in the minor is only one dependency of the group, instead of all dependencies of the group

**Reply by rarkins** [reply]
Look at the branch names and you can see they are grouped as expected. If there's only one upgrade in a group branch then the title switches to the single dependency.

**Reply by ikercrg** [reply]
Ok, now I get it. Thanks for quick answer
